src/hawkes/hawkes_process.jl

```diff
@@ -0,0 +1,169 @@
+function Base.rand(rng::AbstractRNG, hp::HawkesProcess, tmin, tmax)
+    sim = simulate_poisson_times(rng, hp.μ, tmin, tmax) # Simulate Poisson process with base rate
+    sim_desc = generate_descendants(rng, sim, tmax, hp.α, hp.ω) # Recursively generates descendants from first events
+    append!(sim, sim_desc)
+    sort!(sim)
+    return History(sim, fill(nothing, length(sim)), tmin, tmax)
+end
+
+"""
+    StatsAPI.fit(rng, ::HawkesProcess, h::History; step_tol::Float64 = 1e-3, max_iter::Int = 1000)
+
+Expectation-Maximization algorithm from [E. Lewis, G. Mohler (2011)](https://api.semanticscholar.org/CorpusID:2105771)).
+The relevant calculations are in page 4, equations 6-13.
+
+Let (t_1 < ... < t_N) be the event times over the interval [0, T). We use the immigrant-descendant representation,
+where immigrants arrive at a constant base rate μ and each each arrival may generate descendants following the
+activation function α exp(-ω(t - t_i)).
+
+The algorithm consists in the following steps:
+1. Start with some initial guess for the parameters μ, ψ, and ω. ψ = α ω is the branching factor.
+2. Calculate λ(t_i; μ, ψ, ω) (`lambda_ts` in the code) using the procedure in [Ozaki (1979)](https://doi.org/10.1007/bf02480272).
+3. For each t_i and each j < i, calculate D_{ij} = P(t_i is a descendant of t_j).
+        D_{ij} = ψ ω exp(-ω(t_i - t_j)) / λ(t_i; μ, ψ, ω)
+    Define D = sum_{j < i} D_{ij} (expected number of descendants) and div = sum_{j < i} (t_i - t_j) D_ij. 
+4. Update the parameters as
+        μ = (N - D) / T
+        ψ = D / N
+        ω = D / div
+5. If convergence criterion is met, return updated parameters, otherwise, back to step 2.
+
+Notice that in the implementation the process is normalized so the average inter-event time is equal to 1 and, 
+therefore, the interval of the process is transformed from T to N. Also, in equation (8) in the paper,
+sum_{i=1}^N p_{ii} = sum_{i=1}^N (1 - sum{j<i} D_{ij}) = N - D.
+"""
+function StatsAPI.fit(
+    rng::AbstractRNG,
+    ::Type{HawkesProcess},
+    h::History;
+    step_tol::Float64=1e-4,
+    max_iter::Int=1000,
+)
+    N = nb_events(h)
+    N == 0 && return HawkesProcess(0.0, 0.0, 0.0)
+    T = duration(h)
+    norm_ts = h.times .* (N / T) # Average inter-event time equal to 1. Numerical stability
+    n_iters = 0
+    step = step_tol + 1.0
+    lambda_ts = zeros(N)
+    # Step 1 - Choose initial guess such that Λ(T) ≈ N. After normalization, T → N
+    μ = 0.2 + (0.6 * rand(rng)) # μ should not be too close to 0 or 1. μ = 1 → the base rate already accounts for all events
+    ψ = (1 - μ) # ψ = α / β is the integral of the activation function. Λ(t) ≈ μt + ψN_t
+    t90 = 0.5 + (1.5 * rand(rng)) # t90 is the time it takes for the activation function to decay by 90%. Set to t90 ∈ [1/2, 2]
+    ω = log(10.0) * t90 # ω is the parameter corresponding to t90.
+    while (step >= step_tol) && (n_iters < max_iter)
+        # Step 2
+        lambda_ts[1] = 0.0
+        @inbounds for i in 2:N
```

Comment:
Never use `@inbounds` unless the bounds have been provably checked beforehand.

My rule of thumb? Never use `@inbounds`, period. As you get more and more confident writing Julia code, you'll get closer to optimal performance, but right now I'd be very surprised if bounds checking took up the bulk of the runtime. Conversely, `@inbounds` introduces the potential for segmentation faults, which is usually not worth the risk.